### PR TITLE
Refactor divert weaver implementation

### PR DESCRIPTION
**Pull request by pchico83**
Signed-off-by: Pablo Chico de Guzman <pchico83@gmail.com>

# Proposed changes

This is a refactor of the current divert implementation based on weaver.
The idea is to make it easier to add the istio-based divert implementation.

The changes are:
- Create a driver interface with a `Deploy` and `Destroy` implementation that gets called from `cmd/deploy`.
- Implement the driver interface for divert on `pkg/divert/weaver`
- Move the weaver-specific logic from `pkg/k8s/diverts` to `pkg/divert/weaver`
- Keep `pkg/k8s/diverts` to define the weaver divert client. This should probably be renamed to `pkg/k8s/weaver`

There aren't logic changes. The only added code is a test to [test the weaver cache initialization](https://github.com/okteto/okteto/compare/pchico83/refactor-divert?expand=1#diff-48ecc037ba62a8ad948488986d05164025c47d28e84473371b53f98c4b36a6f1)

### CODE REVIEW

**Review by ifbyol — APPROVED** [review]
Minor comments about some pieces of the code but in general looks good. None of the comments is blocker to merge the PR

File: pkg/cmd/stack/divert.go [hunk 1]
```diff
@@ -48,7 +48,7 @@ func applyDivertToService(s *apiv1.Service, old *apiv1.Service) {
 		s.Annotations = map[string]string{}
 	}
 	s.Annotations[model.OktetoDivertServiceAnnotation] = old.Annotations[model.OktetoDivertServiceAnnotation]
-	divertMapping := diverts.PortMapping{}
+	divertMapping := weaver.PortMapping{}
```
[the remark this answers is not in the document]
Comment (pchico83) [reply pkg/cmd/stack/divert.go:51]: yes, it's weaver specific. I will do an interface here and keep it empty for other drivers

File: pkg/divert/weaver/crd.go [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+	"github.com/okteto/okteto/pkg/format"
+	"github.com/okteto/okteto/pkg/k8s/diverts"
+	oktetoLog "github.com/okteto/okteto/pkg/log"
+	"github.com/okteto/okteto/pkg/model"
+	k8sErrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func (d *driver) createDivertCRD(ctx context.Context) error {
+	dClient, err := diverts.GetDivertClient()
```
[the remark this answers is not in the document]
Comment (pchico83) [reply pkg/divert/weaver/crd.go:30]: Great point, I will update it. :-)

File: pkg/divert/weaver/divert.go [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+// PortMapping represents the port mapping of a divert
+type PortMapping struct {
+	ProxyPort          int32 `json:"proxy_port,omitempty" yaml:"proxy_port,omitempty"`
+	OriginalPort       int32 `json:"original_port,omitempty" yaml:"original_port,omitempty"`
+	OriginalTargetPort int32 `json:"original_target_port,omitempty" yaml:"original_target_port,omitempty"`
+}
+
+type driver struct {
+	c     kubernetes.Interface
+	m     *model.Manifest
```
[the remark this answers is not in the document]
Comment (pchico83) [reply pkg/divert/weaver/divert.go:45]: Yes, I will use `client` and `manifest`

**Review by pchico83 — COMMENTED** [review]
Thanks @ifbyol for the comments, I will update the PR accordingly!

**Pull request merged**